wholegames-vcr.Rmd

```diff
@@ -0,0 +1,300 @@
+
+One could imagine other strategies:
+
+* Always having one continuous integration build with vcr turned off but skipping it in contexts where there isn't any token (pull requests from forks for instance?);
+* Only running tests with vcr turned off locally once in a while.
+
+## Summary
+
+* We set up vcr usage in our package exemplighratia by running `use_vcr()` and tweaking the setup file to protect our secret API key and to fool our own package that needs an API token.
+* Inside `test_that()` blocks, we wrapped parts of the code into `vcr::use_cassette()` and ran the tests a first time to generate mock files that hold all information about the API interactions.
+* In one of the tests, we used webmockr to create an environment where only fake requests are allowed. We defined that the request that `gh_organizations()` makes should get a 502 status. We were therefore able to test for the error message `gh_organizations()` returns in such cases.
+
+Now, how do we make sure this works?
+
+* Turn off wifi, run the tests again. It works! Turn on wifi again.
+* Open .Renviron (`usethis::edit_r_environ()`), edit "GITHUB_PAT" into "byeGITHUB_PAT", re-start R, run the tests again. It works! Fix your "GITHUB_PAT" token in .Renviron.
+
+So we now have tests that no longer rely on an internet connection nor on having API credentials.
+
+We also added a continuous integration workflow for having a build using real interactions once every week, as it is important to regularly make sure the package still works against the latest API responses.
+
+For the full list of changes applied to exemplighratia in this chapter, see [the pull request diff on GitHub](https://github.com/maelle/exemplighratia/pull/2/files).
+
+::: {.alert .alert-dismissible .alert-primary} 
+How do we get there with other packages? Let's try httptest in the next chapter!
+:::
+
+## PS: Where to put use_cassette()
+
+Where do we put the `vcr::use_cassette()` call?
+Well, as written in the manual page of that function, _There's a few ways to get correct line numbers for failed tests and one way to not get correct line numbers:_
+What's correct?
+
+* Wrapping the whole `testthat::test_that()` call;
+
+```r
+vcr::use_cassette("thing", {
+  testthat::test_that("thing", {
+    lala <- get_foo()
+    expect_true(lala)
+  })
+})
+```
+
+* Wrapping a few lines inside `testthat::test_that()` **excluding the expectactions `expect_blabla()`**
+
+````r
+testthat::test_that("thing", {
+  vcr::use_cassette("thing", {
+    lala <- get_foo()
+  })
+    expect_true(lala)
+})
+````
+
+What's incorrect?
+
+````r
+testthat::test_that("thing", {
+  vcr::use_cassette("thing", {
+    lala <- get_foo()
+    expect_true(lala)
+  })
+})
+````
+
+We used the solution of only wrapping the lines containing API calls in `vcr::use_cassette()` because that's what vcr maintainer prefers.
```

Comment:
you can remove "because that's what vcr maintainer prefers.", and just say it's up to user, or similar